Async rehydrate with index!

raymondjacobson opened the pull request:
* Adds a migration that adds two columns to the files table
 -- fileName: the actual queryable source filename stripped of any prefixes
 -- dirMultihash: if the file is in an IPFS directory, the CID/multihash for the parent dir

* Updates the models accordingly ^ and updates the image upload flow to populate those two columns

* Allows fallback fetching from FS in /ipfs/:dirCID/:filename endpoint (we previously didn't have this)

* Adds a GET query param to /ipfs/:CID and /ipfs/:dirCID/:filename `fromFS` that async dispatches IPFS rehydration and immediately returns query results from filesystem

* Adds an index to the dirMultihash column

SidSethi commented on creator-node/src/routes/files.js:
```diff
@@ -247,46 +254,97 @@ module.exports = function (app) {
+      dirMultihash: dirCID,
+      // All other file types are valid and can be served through this route.
+      type: { [models.Sequelize.Op.ne]: 'dir' } // Op.ne = notequal
+      fileName: filename
```
this code breaks i believe bc line 262 needs a trailing comma

raymondjacobson replied:
thx! fixed

SidSethi commented on creator-node/sequelize/migrations/20200623002237-add-parent-file-data.js:
```diff
@@ -0,0 +1,85 @@
+'use strict'
+
```
can you make sure to run this migration against a subset of prod data locally and confirm that it doesn't corrupt state? or at least make sure there are files of every "type" locally when you run this migration.

raymondjacobson replied:
Run on staging data locally w/ files of every "type"

SidSethi commented on creator-node/sequelize/migrations/20200623002237-add-parent-file-data.js:
```diff
@@ -0,0 +1,85 @@
+      for (let file of files) {
+        // Set the filename to the last uri "part" of the source file if present
+        // otherwise, use just the sourceFile
+        const fileName = file.sourceFile
```
tested this and p sure it works on all sourceFile formats, but can you add a comment to that effect.
for sourceFile = '....mp3' it matches and returns full string.
what's purpose of the `|| file.sourceFile` though?

raymondjacobson replied:
Added a comment

SidSethi commented on creator-node/sequelize/migrations/20200623002237-add-parent-file-data.js:
```diff
@@ -0,0 +1,85 @@
+        let dirMultihash
+        if (file.type === 'image' && file.storagePath) {
+          const storagePathParts = file.storagePath.split('/')
+          if (storagePathParts.length === 7) {
```
when is the length ever = 7? do you mean 4 here?
i dont think this works?

raymondjacobson replied:
FWIW -- looks like it's different locally vs. in the containers on staging/prod, which is why i had 7. needs another look obv, but separating the migration. though i think we should do it soon

SidSethi commented on creator-node/sequelize/migrations/20200623002237-add-parent-file-data.js:
```diff
@@ -0,0 +1,85 @@
+          }
+        }
+        // Coerce all falsey types to `null`
+        if (!dirMultihash) dirMultihash = null
```
when does this case happen?

raymondjacobson replied:
Added a comment

SidSethi commented on creator-node/src/models/file.js:
```diff
@@ -27,6 +27,16 @@ module.exports = (sequelize, DataTypes) => {
       type: DataTypes.TEXT,
       allowNull: true
     },
+    // the queryable filename if this file entry is inside a dir
```
maybe clarify in comment this is only for files of type "image" else null

raymondjacobson replied:
Added a comment

SidSethi commented on creator-node/src/routes/files.js:
```diff
@@ -105,6 +105,8 @@ module.exports = function (app) {
           multihash: fileResp.hash,
           sourceFile: fileResp.path,
+          dirMultihash: dirCID,
+          fileName: fileResp.path.split('/').slice(-1)[0] || fileResp.path,
```
when is this `|| fileResp.path` case hit? first clause should always be truthy no?

SidSethi commented on creator-node/src/routes/files.js:
```diff
@@ -247,46 +254,97 @@ module.exports = function (app) {
-      multihash: dirCID,
-      type: 'dir'
+      dirMultihash: dirCID,
+      // All other file types are valid and can be served through this route.
```
is this true? this route was originally intended only for multi res images since other tyeps (tracks, metadata) don't have filenames. i think this might work since filename would be = sourceFile for all other types now, but not sure if this is what you intend. you'd have to pass dirCID as empty string / null to get that to work. also below code to parse parentStoragePath would break for non-dir images p sure

raymondjacobson replied:
not sure if i understand this comment. this is the *dir* CID path, but the query is querying for the "leaf" file. Copied comment is weird though, you're right. Cleaning that up & adding a new comment.

SidSethi commented on creator-node/src/routes/files.js:
```diff
@@ -247,46 +254,97 @@ module.exports = function (app) {
+        errorResponseNotFound(`No valid file found for provided dirCID: ${dirCID} and filename: ${filename}`)
+      )
     }
+    const parentStoragePath = queryResults.storagePath.split('/').slice(0, -1).join('/')
```
can you add a comment as to what this is doing. also can we stick to calling it dirStoragePath?

SidSethi commented on creator-node/src/utils.js:
```diff
@@ -132,7 +132,7 @@ async function rehydrateIpfsFromFsIfNecessary (req, multihash, storagePath, file
     let findOriginalFileQuery = await models.File.findAll({
       where: {
-        storagePath: { [models.Sequelize.Op.like]: `%${multihash}%` },
+        dirMultihash: multihash,
```
is this the big perf improvement here? can discuss separately, but per your slack comment with an index on the storagePath column there's not much perf diff between these two versions?

dmanjunath replied:
can't seem to add a comment there, but also change on line 169?

vicky-g commented on creator-node/src/routes/files.js:
```diff
@@ -174,69 +181,69 @@ module.exports = function (app) {
+        )
+      } catch (e) {
+        // If rehydrate throws error, return 500 without attempting to stream file.
+        return sendResponse(req, res, errorResponseServerError(e.message))
```
if rehydration fails, shouldn't we still try to stream the song from fs? / why not attempt to try

raymondjacobson replied:
this is a good Q, but also what we were doing before (i didn't change this aspect of it. maybe we do want to though)

vicky-g replied:
hmm yeah, @dmanjunath was there a reason why we did this? looks like most of the errors are caught in rehydrateIpfsFromFsIfNecessary() anyway so i dont think it'll really hit the catch block here

vicky-g commented on creator-node/src/routes/files.js:
```diff
@@ -174,69 +181,69 @@ module.exports = function (app) {
+      // Retrieves the file directly from the filesystem rather than checking IPFS first
+
+      // Fire-and-forget rehydration
+      rehydrateIpfsFromFsIfNecessary(
```
jc, if this throws an error, will that fail this query?

vicky-g replied:
also it looks like we rehydrate in both if/else cases. should we just always rehydrate if we hit this route?

raymondjacobson replied:
one is awaited, the other is not, there's an important distinction -- in this case here, it will call the method and return the result immediately

raymondjacobson commented:
> overall looks great. thanks for putting this together. had a few questions.
> but also - unless i'm misreading, i dont' think this works? specifically the `storagePath.split('/').length === 7` check should always be false?
> 
> also just want to make sure this is thoroughly tested not just with clean state locally but against existing data, to make sure that (1) migration does not corrupt non multires-image data and (2) new CID routes do not introduce regressions in any of the code paths

I rewrote the split('/')[7] thing that gets the "leaf" of the path as a regex with capture so it's more reliable. I think that should hopefully mitigate some concerns. Tested this against staging data. Migration rollback is pretty bulletproof though so I'm not exceedingly worried about anything.

dmanjunath commented on creator-node/sequelize/migrations/20200623002237-add-parent-file-data.js:
```diff
@@ -0,0 +1,93 @@
+        // Set the dir multihash to the parent CID if there is one
+        let dirMultihash
+        if (file.type === 'image' && file.storagePath) {
+          const match = file.storagePath.match(/(?<parent>Qm.*)\/(?<child>Qm.*)/)
```
i :heart: named capture groups

dmanjunath commented on creator-node/sequelize/migrations/20200623002237-add-parent-file-data.js:
```diff
@@ -0,0 +1,93 @@
+        { transaction }
+      )
+
+      const files = (await queryInterface.sequelize.query(`SELECT * FROM "Files";`, { transaction }))[0]
```
actually pretty sure this isn't going to work. this most likely exceeds node heap size, not to mention running a raw select * will take forever cause these tables have millions of records. we need to do this in a paginated way. like first get count, iterate over chunks of like 50k files. this query seems to work well

```sql
SELECT * FROM "Files" ORDER BY "multihash" ASC LIMIT 100000 OFFSET 200000;
```

raymondjacobson replied:
Updated per our convo offline. Left the bulk update in as a comment.

dmanjunath commented on creator-node/src/models/file.js:
```diff
@@ -38,7 +49,16 @@ module.exports = (sequelize, DataTypes) => {
+  }, {
+    indexes: [
+      {
+        fields: ['multihash']
```
yes confirmed this was already there, and so is cnodeUserUUID

raymondjacobson replied:
added that here as well, we should probs keep the model file up to date

raymondjacobson commented:
Pulled out the route changes in this PR, going in separately. Merging this now.
